students/lee_deitesfeld/lesson04/mailroom.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+
+import sys
+from operator import itemgetter
+
+main_prompt = "\n".join(("Please choose from below mailroom options:",
+          "1 - Manage Donors and Donations",
+          "2 - Create a Report",
+          "3 - Send letters to all donors",
+          "4 - Quit",
+          ">>> "))
+
+
+sub_prompt = "\n".join(("Please make a selection from the following options",
+                        "1 - List the Names of Donors",
+                        "2 - Add a New Donation",
+                        "3 - Return to Main Menu"
+                        ">>> "))
+
+dic_donors = {'donor_db':
+                [
+                {"name": "Bob Jones",
+                 "donations": [63.00, 30.00]
+                },
+                {"name": "Sue Smith",
+                 "donations": [50.00]
+                },
+                {"name": "Joe Grimes",
+                 "donations": [102.00, 45.00, 70.00]
+                },
+                {"name": "Andrea Funk",
+                 "donations": [21.00, 21.00]
+                },
+                {"name": "Mickey Mouse",
+                 "donations": [500.00, 250.00, 100.00]
+                }
+                ]}
+
+def menu_selection(prompt, main_dispatch):
+    while True:
+        response = input(prompt)
+        if main_dispatch[response]() == "exit menu":
+            break
+
+def sub_menu():
+    menu_selection(sub_prompt, sub_dispatch)
+
+def quit():
+    print("Quitting this menu")
+    return "exit menu"
+
+def report():
+    '''Draws up report of names and donations'''
+
+    #calculates sum of donations and adds new key-value pair to donor_db dict
+    for x in dic_donors['donor_db']:
+        ttl_amt = sum(x['donations'])
+        x['sum'] = ttl_amt
+
+    #calculates number of donations and adds new key-value pair to donor_db dict
+    for x in dic_donors['donor_db']:
+        numofdonations = len(x['donations'])
+        x['count'] = numofdonations
+
+    #calculates average of donations and adds new key-value pair to donor_db dict
+    for x in dic_donors['donor_db']:
+        average = sum(x['donations'])/len(x['donations'])
+        str_average = f'{average:.2f}' #limits average to two decimal points
+        x['average'] = str_average
+
+    #sorts list of donors on total amount donated
+    for x in dic_donors['donor_db']:
+        int_ttl_amt = int(x['sum']) #convert ttlamt to integer so sort will work on entire digit (not just first)
+        x['sum'] = int_ttl_amt
+    dic_donors['donor_db'].sort(key=itemgetter('sum'), reverse = True) #sort on key='sum'
+    for x in dic_donors['donor_db']:
+        ttl_amt_2dec = '%.2f' % x['sum'] #limits total to two decimal points
+        x['sum'] = ttl_amt_2dec
+
+    #removes history of donations, since it's not in final report
+    for x in dic_donors['donor_db']:
+        del x['donations']
+
```

Comment:
you have 6 loops above when it seems you can have just one loop since you're just looping over the same data structure?

also, I would avoid modifying the main data structure in place - for you report you should create a new data structure to hold your report data.